shared/Microsoft.AspNetCore.Certificates.Generation.Sources/CertificateManager.cs

```diff
@@ -387,34 +402,181 @@ public bool IsTrusted(X509Certificate2 certificate)
             {
                 store.Open(OpenFlags.ReadWrite);
-                foreach (var certificate in certificates)
+                var matching = store.Certificates
+                    .OfType<X509Certificate2>()
+                    .Single(c => c.SerialNumber == certificate.SerialNumber);
+
+                store.Remove(matching);
+                store.Close();
+            }
+        }
+
+        private void RemoveCertificateFromTrustedRoots(X509Certificate2 certificate)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                using (var store = new X509Store(StoreName.Root, StoreLocation.CurrentUser))
                 {
-                    var matching = store.Certificates.OfType<X509Certificate2>().Single(c => c.SerialNumber == certificate.SerialNumber);
+                    store.Open(OpenFlags.ReadWrite);
+                    var matching = store.Certificates
+                        .OfType<X509Certificate2>()
+                        .Single(c => c.SerialNumber == certificate.SerialNumber);
+
                     store.Remove(matching);
+                    store.Close();
+                }
+            }
+            else
+            {
+                if (IsTrusted(certificate)) // On OSX this check just ensures its on the system keychain
+                {
+                    try
+                    {
+                        RemoveCertificateTrustRule(certificate);
+                    }
+                    catch
+                    {
+                        // We don't care if we fail to remove the trust rule if
+                        // for some reason the certificate became untrusted.
+                        // The delete command will fail if the certificate is
+                        // trusted.
+                    }
+                    RemoveCertificateFromKeyChain(MacOSSystemKeyChain, certificate);
                 }
-                store.Close();
+            }
+        }
+
+        private static void RemoveCertificateTrustRule(X509Certificate2 certificate)
+        {
+            var certificatePath = Path.GetTempFileName();
+            try
+            {
+                var certBytes = certificate.Export(X509ContentType.Cert);
+                File.WriteAllBytes(certificatePath, certBytes);
+                var processInfo = new ProcessStartInfo(
+                    MacOSRemoveCertificateTrustCommandLine,
+                    string.Format(
+                        MacOSRemoveCertificateTrustCommandLineArgumentsFormat,
+                        certificatePath
+                    ))
+                {
+                    RedirectStandardOutput = true,
+                    RedirectStandardError = true
+                };
+                var process = Process.Start(processInfo);
+                process.WaitForExit();
+                var output = process.StandardOutput.ReadToEnd() + process.StandardError.ReadToEnd();
```

Comment:
You should read the output/error before WaitForExit() - as that can result in a deadlock: https://msdn.microsoft.com/en-us/library/system.diagnostics.process.standardoutput(v=vs.110).aspx

Reply:
Yeah, for some reason I thought it was the other way around.